csharp/src/Ice/Connection.cs

```diff
@@ -222,17 +224,35 @@ public ConnectionInfo GetConnectionInfo()
         }
 
         /// <summary>Send a heartbeat message.</summary>
-        public void Heartbeat() => HeartbeatAsync().Wait();
+        public void Heartbeat()
+        {
+            try
+            {
+                HeartbeatAsync().AsTask().Wait();
+            }
+            catch (AggregateException ex)
+            {
+                Debug.Assert(ex.InnerException != null);
+                throw ex.InnerException;
+            }
+        }
 
         /// <summary>Send an asynchronous heartbeat message.</summary>
         /// <param name="progress">Sent progress provider.</param>
         /// <param name="cancel">A cancellation token that receives the cancellation requests.</param>
-        public Task HeartbeatAsync(IProgress<bool>? progress = null, CancellationToken cancel = new CancellationToken())
+        public async ValueTask HeartbeatAsync(IProgress<bool>? progress = null, CancellationToken cancel = default)
         {
-            var completed = new HeartbeatTaskCompletionCallback(progress, cancel);
-            var outgoing = new HeartbeatOutgoingAsync(this, _communicator, completed);
-            outgoing.Invoke();
-            return completed.Task;
+            Task writeTask;
+            lock (_mutex)
+            {
+                if (_state >= State.Closed)
+                {
+                    throw _exception!;
+                }
+                writeTask = SendFrameAsync(() => GetProtocolFrameData(_validateConnectionFrame));
```

Comment:
We could pass progress... but what would this buy us? Doing it from the proxy `InvokeAsync` method seems simpler to me unless of course we change the semantics of `SendRequestAsync` to also wait for the response.

As for the cancelation token, are you suggesting that we actually cancel the transport send when the invocation is canceled? This would cause the other pending requests waiting for sending to be canceled as well.